constraint-solver/src/solver/bus_interaction_variable_wrapper.rs

```diff
@@ -28,126 +28,106 @@ impl<V: Display> Display for Variable<V> {
-                            expr.transform_var_type(&mut |v| Variable::Variable(v.clone()));
-                        let v = Variable::BusInteractionField(bus_interaction_index, field_index);
-                        new_constraints.push(
-                            transformed_expr - GroupedExpression::from_unknown_variable(v.clone()),
-                        );
-                        bus_interaction_vars.insert(v.clone(), expr.clone());
-                        GroupedExpression::from_unknown_variable(v)
-                    },
-                ))
-            })
-            .collect();
-        let constraint_system = ConstraintSystem {
-            algebraic_constraints: constraint_system
-                .algebraic_constraints
-                .iter()
-                .map(|expr| expr.transform_var_type(&mut |v| Variable::Variable(v.clone())))
-                .chain(new_constraints)
-                .collect(),
-            bus_interactions,
-        };
-        (
-            Self {
-                bus_interaction_vars,
-            },
-            constraint_system,
-        )
-    }
-
-    pub fn finalize(
-        mut self,
-        solve_result: SolveResult<T::Transformed, Variable<V>>,
-    ) -> SolveResult<T, V> {
-        for (variable, expr) in &solve_result.assignments {
-            // Apply any assignments to the bus interaction field definitions.
-            if let Variable::BusInteractionField(..) = variable {
-                // Non-concrete assignments are only generated by the `quadratic_equivalences` module,
-                // and bus interaction fields should only appear in bus interactions and constraints
-                // of the form <bus interaction field> = <expression>, which would not lead to a
-                // quadratic equivalence.
-                // So we only expect concrete values here, but we need to do some trickery
-                // to translate types which we expect to be equal anyway.
-                let value: T::FieldType = expr
-                    .try_to_known()
+    let mut result = Vec::<Substitution<T, V>>::new();
+    for (variable, expr) in &assignments {
+        // Apply any assignments to the bus interaction field definitions.
+        if let Variable::BusInteractionField(..) = variable {
+            // Non-concrete assignments are only generated by the `quadratic_equivalences` module,
+            // and bus interaction fields should only appear in bus interactions and constraints
+            // of the form <bus interaction field> = <expression>, which would not lead to a
+            // quadratic equivalence.
+            // So we only expect concrete values here, but we need to do some trickery
+            // to translate types which we expect to be equal anyway.
+            let value = GroupedExpression::from_number(
+                expr.try_to_known()
                     .unwrap()
                     .transform_var_type(&mut |_| unreachable!())
```

Comment:
This line is not needed, is it? If there was a variable, the following `.try_to_number().unwrap()` would fail.